src/helpers/buffers/unordered_receiver.rs

```diff
@@ -0,0 +1,502 @@
+    /// they are stashed.  These are more than `c` items past the current
+    /// item into the future when registered (c = capacity or `wakers.len()`).
+    /// So we don't want to wake them frequently.  Instead, these are woken on
+    /// a fixed cadence of every `c/2` items.
+    ///
+    /// When polled again, any that were only a little past the capacity will
+    /// be entered into normal `wakers` correctly.  Those that are too far
+    /// ahead (i.e., `d = i - (next + c)` is large) they will be woken at most
+    /// `1 + 2d/c` times extra.
+    ///
+    /// Assuming that the awoken items are polled in a timely fashion, this
+    /// ensures that any overflow will be registered in `wakers` (or read)
+    /// before the data needs to be read.  Notifying every `c` items rather
+    /// than `c/2` could mean that a task is not able to poll and enter
+    /// `wakers` in time to be read out.
+    ///
+    /// Note: in protocols we try to send before receiving, so we can rely on
+    /// that easing load on this mechanism.  There might also need to be some
+    /// end-to-end back pressure for tasks that do not involve sending at all.
+    overflow_wakers: Vec<Waker>,
+    /// Left over data from a previous `recv()` call.  The underlying stream
+    /// can provide chunks of data larger than a single message.  We save the
+    /// spare data here.
+    spare: Spare,
+    /// The absolute index of the next value that will be received.
+    next: usize,
+    _marker: PhantomData<C>,
+}
+
+impl<S, C> OperatingState<S, C>
+where
+    S: Stream<Item = C>,
+    C: AsRef<[u8]>,
+{
+    /// Determine whether `i` is the next record that we expect to receive.
+    fn is_next(&self, i: usize) -> bool {
+        i == self.next
+    }
+
+    /// Track a waker from a future that was invoked before data was ready.
+    ///
+    /// Note that this only tracks the last waker for each index, which appears
+    /// to violate the contract for `Future` implementations.  That shouldn't a
+    /// problem in practice because we expect [`recv`] to be called only once,
+    /// which produces just one `Future`.  Informing the last context to [`poll`]
+    /// that `Future` should at least ensure that things progress, even if we
+    /// don't guarantee that all contexts are awoken.
+    ///
+    /// # Panics
+    ///
+    /// If `i` is for an message that has already been read.
+    ///
+    /// [`recv`]: UnorderedReceiver::recv
+    /// [`poll`]: Future::poll
+    fn add_waker(&mut self, i: usize, waker: Waker) {
+        // We don't save a waker at `self.next`, so `>` and not `>=`.
+        if i > self.next + self.wakers.len() {
+            self.overflow_wakers.push(waker);
+        } else {
+            assert!(
+                i > self.next,
+                "Awaiting a read that has already been fulfilled"
+            );
+            let index = i % self.wakers.len();
+            self.wakers[index] = Some(waker);
```

Comment:
maybe another debug assert that `self.wakers[index] == None`?

Reply:
That's not going to be the case.  I originally had that and found that there are some cases (the synchronous test is one), where we replace a waker.  Basically, if the future is polled more than once in this state, we will be getting another waker for the same item.  I think that a strict interpretation of the `Future` contract requires us to store both wakers in that case, but the note above explains why I think that's unnecessary.

(I spent a lot of time thinking about this and I'm somewhat sure that this is not OK in theory, but maybe OK in practice.  That is, what if the first to poll is kept, the second is polled and discarded? We won't wake the first.  But then, we won't be doing that, I hope.)

Reply:
I've got a check for this that should ensure that it is robust.  Not guaranteed (`Waker::will_wake` can return false in cases where there is no issue), but then we shouldn't be hitting that code path outside of the few tests I have.